planner/core/fragment.go

```diff
@@ -59,17 +59,48 @@ func (e *mppTaskGenerator) generateMPPTasks(s *PhysicalExchangeSender) ([]*kv.MP
 		StartTs: e.startTS,
 		ID:      -1,
 	}
-	rootTasks, err := e.generateMPPTasksForFragment(s.Fragment)
+	rootTasks, err := e.generateMPPTasksForFragment(s)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	s.TargetTasks = []*kv.MPPTask{tidbTask}
 	return rootTasks, nil
 }
 
-func (e *mppTaskGenerator) generateMPPTasksForFragment(f *Fragment) (tasks []*kv.MPPTask, err error) {
+func (f *Fragment) init(p PhysicalPlan) error {
+	if ts, ok := p.(*PhysicalTableScan); ok {
+		if f.TableScan != nil {
+			return errors.New("one task contains at most one table scan")
+		}
+		f.TableScan = ts
+		return nil
+	}
+	if recv, ok := p.(*PhysicalExchangeReceiver); ok {
+		f.ExchangeReceivers = append(f.ExchangeReceivers, recv)
+		return nil
+	}
+	for _, ch := range p.Children() {
+		if err := f.init(ch); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
```

Comment:
use switch ... case... ?

Reply:
There are only two branches. I think using switch is a little waste.

Reply:
indeed tree, including the default